sdk/identity/azure-identity/src/main/java/com/azure/identity/SharedTokenCacheCredential.java

```diff
@@ -71,33 +73,25 @@
         // find if the Public Client app with the requested username exists
         return Mono.fromFuture(pubClient.getAccounts())
             .flatMap(set -> {
-                IAccount requestedAccount = null;
+                IAccount requestedAccount;
+                Map<String, IAccount> accounts = new HashMap<>(); // home account id -> account
 
-                if (username != null) {
-                    for (IAccount cached : set) {
-                        if (username.equals(cached.username())) {
-                            requestedAccount = cached;
-                            break;
-                        }
-                    }
-                } else {
-                    for (IAccount cached : set) {
-                        if (requestedAccount != null) {
-                            if (!cached.username().equals(requestedAccount.username())
-                                    || !cached.environment().equals(requestedAccount.environment())
-                                    || !cached.homeAccountId().equals(requestedAccount.homeAccountId())) {
-                                return Mono.error(new RuntimeException("Multiple accounts found in the token cache. "
-                                    + "Please specify the username of the account in the AZURE_USERNAME environment "
-                                    + "variable"));
-                            }
-                        } else {
-                            requestedAccount = cached;
+                for (IAccount cached : set) {
+                    if (username == null || username.equals(cached.username())) {
+                        if (!accounts.containsKey(cached.homeAccountId())) {
+                            accounts.put(cached.homeAccountId(), cached);
                         }
                     }
                 }
 
-                if (requestedAccount == null) {
+                if (accounts.size() == 0) {
                     return Mono.error(new RuntimeException("Requested account was not found"));
```

Comment:
Seems like this could also be hit if there are no accounts no?

Reply:
Added better message